virtcontainers/clh.go

```diff
@@ -0,0 +1,1260 @@
+
+	"github.com/kata-containers/runtime/virtcontainers/device/config"
+	"github.com/kata-containers/runtime/virtcontainers/store"
+	"github.com/kata-containers/runtime/virtcontainers/types"
+	"github.com/kata-containers/runtime/virtcontainers/utils"
+)
+
+//
+// Constants and type definitions related to cloud hypervisor
+//
+
+type clhState uint8
+
+const (
+	clhNotReady clhState = iota
+	clhReady
+)
+
+const (
+	clhTimeout            = 10
+	clhSocket             = "clh.sock"
+	vfsdSocket            = "virtiofsd.sock"
+	clhSerial             = "serial-tty.log"
+	clhStopSandboxTimeout = 15
+	supportedMajorVersion = 0
+	supportedMinorVersion = 3
+	defaultIchPath        = "/usr/local/bin/cloud-hypervisor"
+)
+
+type CloudHypervisorVersion struct {
+	Major    int
+	Minor    int
+	Revision int
+}
+
+//
+// Cloud hypervisor state
+//
+type CloudHypervisorState struct {
+	state        clhState
+	PID          int
+	VirtiofsdPID int
+}
+
+type cloudHypervisor struct {
+	id         string
+	state      CloudHypervisorState
+	clhd       *exec.Cmd
+	store      *store.VCStore
+	config     HypervisorConfig
+	ctx        context.Context
+	socketPath string
+	version    CloudHypervisorVersion
+	cliBuilder *DefaultCLIBuilder
+}
+
+var clhKernelParams = []Param{
+
+	{"root", "/dev/vda1"},
+	{"panic", "1"},              // upon kernel panic wait 1 second before reboot
+	{"no_timer_check", ""},      // do not check broken timer IRQ resources
+	{"noreplace-smp", ""},       // do not replace SMP instructions
+	{"page_alloc.shuffle", "1"}, // enable page allocation shuffling
+}
+
+var clhDebugKernelParams = []Param{
+
+	{"console", "ttyS0,115200n8"},     // enable serial console
+	{"systemd.log_level", "debug"},    // enable systemd debug output
+    {"systemd.log_target", "console"}, // send loggng to the console
+	{"initcall_debug", "1"},           // print init call timing information to the console
+}
+
+//###########################################################
+//
+// hypervisor interface implementation for cloud-hypervisor
+//
+//###########################################################
+
+func (clh *cloudHypervisor) createSandbox(ctx context.Context, id string, networkNS NetworkNamespace, hypervisorConfig *HypervisorConfig, vcStore *store.VCStore) error {
+	clh.ctx = ctx
+
+	span, _ := clh.trace("createSandbox")
+	defer span.Finish()
+
+	err := hypervisorConfig.valid()
+	if err != nil {
+		return err
+	}
+
+	clh.id = id
+	clh.store = vcStore
+	clh.config = *hypervisorConfig
+	clh.state.state = clhNotReady
+
+	err = clh.getAvailableVersion()
+	if err != nil {
+		return err
+	}
+
+	if clh.version.Major < supportedMajorVersion && clh.version.Minor < supportedMinorVersion {
+		errorMessage := fmt.Sprintf("Unsupported version: cloud-hypervisor %d.%d not supported by this driver version (%d.%d)",
+			clh.version.Major,
+			clh.version.Minor,
+			supportedMajorVersion,
+			supportedMinorVersion)
+		return errors.New(errorMessage)
+	}
```

Comment:
I'd rather not doing those checks. The support is going to be experimental at first anyways, and until the CLI and HTTP API stabilize, I don't think we should bother with this kind of check.

Reply:
But it is a fact that any binary below 0.3.0 the driver will fail to drive due to changes in the cli. The smallest glitch in the cli args and the hypervisor will not start and this is very difficult to troubleshoot if you do not know where to look for. I would actually like to see a different versioning for both cli and http api which the --version flag prints on seperate lines. This way we can instantiate different builders or command patterns for cli and http based on those versions.

Reply:
> But it is a fact that any binary below 0.3.0 the driver will fail to drive due to changes in the cli. 

I agree. I don't have a strong opinion on that one, we can keep it.

> I would actually like to see a different versioning for both cli and http api

That one I disagree with. It's ok to break the CLI/API between minor versions, especially with such a new project, but the CLI and HTTP APIs should be kept in sync when they map. There will be HTTP API endpoints that don't have a CLI counterpart (e.g. the hotplug ones), but for those who do they should be consistent.